source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp

```diff
@@ -0,0 +1,259 @@
+  this->points_.clear();
+  this->times_.clear();
+}
+
+template<typename TrajectoryT>
+template<typename DurationT>
+inline void TrajectoryBase<TrajectoryT>::add_point(
+    const TrajectoryT& new_point, const std::chrono::duration<int64_t, DurationT>& new_time) {
+  this->set_empty(false);
+  this->points_.push_back(new_point);
+
+  if (!this->times_.empty()) {
+    auto const previous_time = this->times_.back();
+    this->times_.push_back(previous_time + new_time);
+  } else {
+    this->times_.push_back(new_time);
+  }
+}
+
+template<typename TrajectoryT>
+template<typename DurationT>
+inline void TrajectoryBase<TrajectoryT>::insert_point(
+    const TrajectoryT& new_point, const std::chrono::duration<int64_t, DurationT>& new_time, int pos) {
+  this->set_empty(false);
+
+  auto it_points = this->points_.begin();
+  auto it_times = this->times_.begin();
+  std::advance(it_points, pos);
+  std::advance(it_times, pos);
+
+  this->points_.insert(it_points, new_point);
+
+  auto previous_time = this->times_[pos - 1];
+  this->times_.insert(it_times, previous_time + new_time);
+
+  for (unsigned int i = pos + 1; i <= this->points_.size(); i++) {
+    this->times_[i] += new_time;
+  }
+}
+
+template<typename TrajectoryT>
+inline void TrajectoryBase<TrajectoryT>::delete_point() {
+  this->set_empty(false);
+  if (!this->points_.empty()) {
+    this->points_.pop_back();
+  }
+  if (!this->times_.empty()) {
+    this->times_.pop_back();
+  }
+}
+
+template<typename TrajectoryT>
+inline void TrajectoryBase<TrajectoryT>::clear() {
+  this->points_.clear();
+  this->times_.clear();
+}
+
+template<typename TrajectoryT>
+inline const std::deque<TrajectoryT>& TrajectoryBase<TrajectoryT>::get_points() const {
+  return this->points_;
+}
+
+template<typename TrajectoryT>
+inline const TrajectoryT& TrajectoryBase<TrajectoryT>::get_point(unsigned int index) const {
+  return this->points_[index];
+}
+
+template<typename TrajectoryT>
+inline TrajectoryT& TrajectoryBase<TrajectoryT>::get_point(unsigned int index) {
+  return this->points_[index];
+}
+
+template<typename TrajectoryT>
+template<typename DurationT>
+inline bool TrajectoryBase<TrajectoryT>::set_point(
+    unsigned int index, const TrajectoryT& point, const std::chrono::duration<int64_t, DurationT>& new_time) {
+  if (index < this->points_.size()) {
+    this->points_[index] = point;
+    if (index == 0) {
+      this->times_[index] = new_time;
+    } else {
+      this->times_[index] = this->times_[index - 1] + new_time;
+    }
+    for (unsigned int i = index + 1; i < this->points_.size(); ++i) {
+      this->times_[i] += this->times_[index - 1];
+    }
+    return true;
+  }
+  return false;
+}
+
+template<typename TrajectoryT>
+template<typename DurationT>
+inline bool TrajectoryBase<TrajectoryT>::set_points(
+    const std::vector<TrajectoryT>& points, const std::vector<std::chrono::duration<int64_t, DurationT>>& new_times) {
+  if (points.size() != new_times.size()) {
+    return false;
```

Comment:
In control libraries we usually throw in these cases, not returning a bool (same above)